MOC Sensitivity setup

timothyas opened the pull request:
Hey there, this is a follow up to MITgcm/MITgcm#405 with my MOC sensitivity setup from Smith & Heimbach 2019. I added
- `code_ad.moc` which only changes EXF_OPTIONS.h, basically allowing one to compute net heat/salt sensitivites while prescribing the atmospheric state
- `input_ad.moc` with the data.ctrl and data.ecco

My questions are:
1. I'm not sure if I did the "prepare_run" correctly - it seemed reasonable to use the same one that the ecco_v4 setup, but let me know if that looks fishy
2. The MOC objective function requires masks to define where it is applied. I have those files, where should I put them?
3. I made this an "_ad" experiment, which seemed sensible, but happy to change that. I'm not sure how that is all handled with taf etc.

jm-c commented:
@timothyas Hi, I talked a little bit with @mjlosch regarding the best place to add this. This llc90 experiment is certainly a nice
place, but with the caveat that this Adjoint set-up requires some computing resources to run and as a result is not testing as often
as others. An alternative would be to add it to "global_oce_cs32", a smaller set-up that compiles and uses the same list
of pacakges as llc90i but run faster and is also tested more often. The question here is that I don't know how much effort it would
require for you to switch it from llc90 to cs32. Would you mind taking a quick look and see if it's reasonable or not ?

jm-c commented:
Forgot to mention, global_oce_cs32 is also in verification_other repos, just to clarify.

timothyas commented:
Thanks @jm-c, that all makes sense to me. The data files actually should not change at all, since it's just specifying the objective function. I would just have to make a mask with the cs32 grid, which would take a little bit of wrangling but shouldn't be too bad (famous last words). I'm up for it. 

Where would said masks go?

jm-c commented:
@timothyas I think you can put this mask(s) in global_oce_cs32/input_ad/ : it will be available to any of the adjoint run 
(symbolic link to any AD run-dir) for this cs32 set-up, but will not be available for the forward only run, which is probably what we want I guess.

timothyas commented:
Now that I'm poking around I think this will be even more simple than expected. I can just use the same masks that are in global_oce_cs32/input_ad.sens/  and simply add the new objective function to that list

timothyas commented:
@jm-c Now implemented:
1. Added the moc objective function to the current ones in global_oce_cs32/input_ad.sens/data.ecco
2. Added a new code directory: global_oce_cs32/code_ad.sens, that links every file to global_oce_cs32/code, except EXF_OPTIONS.h, which has the flags allowing the user to compute sensitivity to hflux/sflux while prescribing atm state
3. Added hflux/sflux to control variables in new input_ad.sens/data.ctrl that uses that capability

mjlosch commented on global_oce_cs32/input_ad.sens/data.ecco:
```diff
@@ -20,6 +20,13 @@
  gencost_outputlevel(2)=1,
  mult_gencost(2) = 1.,
+#
+ gencost_avgperiod(1)  = 'step',
```
Shouldn't this be gencost_avgperiod(3), etc.?

mjlosch replied:
@timothyas
 One more thing, I need to change `wt_ones.data` to `wt_ones.bin` in data.ctrl to make it work and then the results is:
```
Y Y Y Y  0> 1< 1 FAIL  global_oce_cs32.sens
```
Is that expected? Probably, because the cost function changes. Maybe @jm-c can provide a reference results from the reference machine. Do you want me to push the `wt_ones.bin` fix?

timothyas replied:
Hey @mjlosch thanks for these catches! I swear sometimes my code works ...😅
I'll go ahead and take care of the wt_ones.bin fix, thanks though

timothyas replied:
As for the fail, I'm not sure. Might have something to do with the MOC code not being merged, but I'll test it right now.

timothyas replied:
Is the test running a forward or adjoint simulation? I just ran this and it failed for the forward run because the grdchk package is turned on, which looks for adxx_* and can't find them. The adjoint run produces NaNs, but runs just fine.

I ran the adjoint both on master and on top of the MITgcm/MITgcm#405 branch, and it worked for both. Except when it ran on the master branch, the run just didn't recognize the MOC objective function.

mjlosch replied:
I didn't get any NaN.

mjlosch reviewed (commented):
Except for the one comment, this looks good.

mjlosch commented:
@timothyas 
Sorry for the confusion with putting this in the wrong thread. I only tried `../verification/testreport -t global_oce_cs32 -adm -ncad -devel -j 4` with MITgcm/MITgcm#405. I am not surprised that the results change, when you change the cost function. But I think we should let @jm-c update the reference result, as he has access to the reference machine.

mjlosch commented:
@timothyas 
sorry about the confusion again. I reran after you latest push and now i get this of the TAF AD test:
```
Y Y Y Y 16>16< 6 pass  global_oce_cs32  (e=0, w=48)
Y Y Y Y 16>16<16 pass  global_oce_cs32.sens
```
Is that expected?

timothyas commented:
Hey @mjlosch, sorry to be slow on this, I was at first just running the setup, now I'm also running testreport. I do not know why I'm struggling so much with this, but on one machine I'm getting NaNs and on the other, the .sens run simply fails once it starts the adjoint part because EXF_INTERP_READ grabs a bad file name - perhaps just dumb compile issues or something. I'll keep hacking. Sorry! I don't know what the deal is. 

Either way, I'm surprised the `.sens` test passes because I would expect it to produce a different cost function and sensitivity, I would've expected a failure, which was what you mentioned first?

timothyas commented:
Ok I made some progress on this finally. It turns out that  MITgcm/MITgcm#405 had a bug in it related to a global sum operation, only applicable to the case of multiple threads per processor. That is fixed with the most recent commit. Without that fix, the MOC objective function was just zero, presumably because it was just grabbing an empty tile, and so the tests shown above were passing because the cost function was no different. I'm wondering if this bad global sum operation was what was giving me NaNs, since I am not getting them anymore.

Now, I am sure that this cs32 test is computing the correct MOC because I've verified the calculation on and offline (costfunction reproducing my python script from reading `m_trVol_step`). However, the MOC objective function is causing the adjoint to fail on the first reverse step with too small rStar. I'm not sure how to resolve this one yet.

Another small issue is that in adjoint mode I can't seem to get the MOC f_gencost values to print to STDOUT or to costfunction0000. Not really a big deal, but if anyone knows off the top of their head that would be bonus, but first the rstar issue...

Thanks for your time - sorry, this has turned out to be more trouble than I thought!

mjlosch commented:
@timothyas 
Now, after finally running this with the correct versions of code and verification experiments (both branches are called "samoc", right?), I am also getting an error and that's related to a tape not properly defined.
```
At line 711 of file cost_gencost_moc_ad.f
Fortran runtime error: Index '2' of dimension 1 of array 'tapelev_init_mytempmax_3h' above upper bound of 1
```
I use `../verification/testreport -t global_oce_cs32 -adm -ncad -devel` and in particular the `-devel` option to turn on array bound checking with gfortran. Exceeding the array boundaries may have caused your NaN?

When you replace `tapelev_init` with a newly defined 
```CADJ  INIT tapelev_moc   = common, Nr```
the corresponding `tapelev_moc_mytempmax_3h` has the correct length (`Nr`). It also greatly simplifies the generated code (no more cost_gencost_mocmd). With this fix, I get this:
```
Y Y Y Y 16>16< 6 pass  global_oce_cs32  (e=0, w=48)
Y Y Y Y  0> 7< 0 FAIL  global_oce_cs32.sens
```
but also:
```
> ** WARNING ** ECCO_CHECK: cost_gencost_transp/moc.F not tested
>  See m_horflux_vol via cost_gencost_boxmean.F.
> ** WARNING ** ECCO_CHECK: cost_gencost_transp/moc.F not tested
>  See m_horflux_vol via cost_gencost_boxmean.F.
```
Is this expected?
Shall I push my fix to  MITgcm/MITgcm#405 ?

timothyas commented:
Thanks so much for your effort @mjlosch. Yes, please push your changes, thanks for that. I guess I still have a lot to learn with regard to the tapes! 

As for the warnings about `transp/moc not tested`: I put these in for the `gencost_transp` cost function a long time ago, something we had agreed upon in MITgcm/MITgcm#102. Now, though they seem irrelevant and I can take them out for the MOC objective function.

timothyas commented:
Wow @mjlosch - nice fix! Now the simulation is no longer crashing with the R_STAR issues (and, the print statements are working in adjoint mode). Amazing.

This was just a regular run, I'm now running with testreport just to double check (the reason I was not doing that is because I've been using our compute nodes on the local cluster, which don't have access to `cc`, which it seems testreport needs, even with the `-runonly` flag, so been running testreport on a desktop that takes much longer). 

Seems good to go, will just confirm that testreport is all good when that finishes later today.

timothyas commented:
Ok I'm getting strange errors when running the compilation with testreport, such as the one shown below. No idea what the issue is here. Have you seen anything like this before?

```
ecco_phys_ad.f(2064): warning #8889: Explicit declaration of the EXTERNAL attribute is required.   [ECCO_DIV]
          call ecco_div( gencost_storefld(1-olx,1-oly,1,1,kgen),1,
---------------^
```

It seems to have a problem with a different subroutine every time I compile.

mjlosch commented:
> Ok I'm getting strange errors when running the compilation with testreport, such as the one shown below. No idea what the issue is here. Have you seen anything like this before?
> 
> ```
> ecco_phys_ad.f(2064): warning #8889: Explicit declaration of the EXTERNAL attribute is required.   [ECCO_DIV]
>           call ecco_div( gencost_storefld(1-olx,1-oly,1,1,kgen),1,
> ---------------^
> ```
No, haven't seen that before. That's strange, because ecco_div is a subroutine and not a function. Functions require an EXTERNAL attribute, but subroutines don't. Something wrong with the compiler?

timothyas commented:
Possibly ... will try on a different machine later ...

timothyas commented:
I think I was getting mixed up with the various gcc/intel modules on my desktop. Got that all sorted, and I'm getting the same thing @mjlosch is with testreport:

```
Y Y Y Y 16>16< 6 pass  global_oce_cs32  (e=0, w=48)
Y Y Y Y  0> 7< 0 FAIL  global_oce_cs32.sens
```

One final thing: the additional lines in `code_ad.sens/EXF_OPTIONS.h`:
```
C- For computing sensitivities to net fluxes with provided ATM state
#define CALC_HFLUX_SENS
#define CALC_SFLUX_SENS
```

are not being tested because the test is using `code_ad`. I think we may as well test it, but what do you guys think? If so, should I just incorporate it into `code_ad/EXF_OPTIONS.h` to avoid compiling twice?

jm-c commented:
@timothyas  Thanks for all the work you did to make this new feature tested. I guess part of it was just to make it to fit into 
"a good regression test" format ; but in the process, I think it also gave you the chance to fix/improve some pieces of the new code.
And regarding the code_ad.sens vs code_ad location, I guess @mjlosch  will take a look at but you are right that only "code_ad" is
used (since only one executable per experiment is built).

timothyas commented:
Sounds good @jm-c. Yeah I appreciate the process that helped improve the code - big thanks to @mjlosch there!

mjlosch commented:
@timothyas I think that this CALC_H/SFLUX_SENS code should be tested. I am not quite sure where to put these flags, or in fact, what to call them. There's already the similar (same) ALLOW_H/SFLUX_CONTROL in the ECCO_DEPRECATED block and these are defined int CTRL_OPTIONS.h.

I wonder if you need these CPP flags at all with the useGenCtrl and xx_gentim2d-machinery. Have you tried to remove them from the code?

timothyas commented:
I can re-test this, but I've basically only used the xx_gentim2d machinery. I added these statements many moons ago when I was starting the samoc project and wanted h/sflux sensitivities with the ECCO setup. I recall that it did not work without the flags, which is why I added the code block in exf_getsurfacefluxes.F.

Albeit, that was with an old checkout of the MITgcm, so I can try removing the flags

mjlosch commented:
@timothyas 
I tested this with removed flags and none of the current AD-verification experiments fail, so I think you can safely remove them and this would then make the extra directory code_ad.sense superfluous.

mjlosch commented:
@timothyas 
just to be clear, the code is necessary, but the #ifdef lines can be removed.

timothyas commented:
So I had made the extra macros to avoid the situation of someone adding an hflux/sflux control variable with prescribed atmospheric state, and this makes it more intentional that someone would want to compute sensitivities. Do you think this is an unnecessary guard rail?

mjlosch commented:
I think that you don't need these macros. They sort of beat the purpose of the generic controls, as far as I can see. One can imagine adding some documentation to chapter 10, table 10.8: 
https://mitgcm.readthedocs.io/en/latest/ocean_state_est/ocean_state_est.html#gencost-ctrl-files

But if you think that they are indeed necessary, then I would move the definition to CTRL_OPTIONS.h

timothyas commented:
I'll just go ahead and remove them. I'll ping you when that's done.

timothyas commented:
@mjlosch 
in MITgcm/MITgcm#405
- [x] EXF_OPTIONS.h macros removed from code and exf documentation
- [x] added xx_h/sflux to gentim2d controls in docs
- [x] added generic note about the generic controls to be mindful of criss crossing control variables

Here:
- [x] `code_ad.sens` removed 
- [ ] rerunning tests

timothyas commented:
Alrighty @mjlosch and @jm-c I think the finish line is in sight, thanks for your patience. Tests ran without a hitch, barring the expected run time value comparison failure: 

```
Y Y Y Y 16>16< 6 pass  global_oce_cs32  (e=0, w=48)
Y Y Y Y  0> 7< 0 FAIL  global_oce_cs32.sens
```

mjlosch commented:
LGTM, let's wait until MITgcm/MITgcm#405 is merged, and then we can give it another test and ask @jm-c to update the reference output.

jm-c commented:
Now that https://github.com/MITgcm/MITgcm/pull/405 has been merged in, will merge this one soon (after updating the output_adm.sens.txt).
